src/templater.rs

```diff
@@ -443,3 +444,16 @@ impl TemplateProperty<Signature, Timestamp> for SignatureTimestamp {
         context.timestamp.clone()
     }
 }
+
+pub struct EmptyProperty<'a> {
+    pub repo: RepoRef<'a>,
+}
+
+impl TemplateProperty<Commit, bool> for EmptyProperty<'_> {
+    fn extract(&self, context: &Commit) -> bool {
+        // true iff the commit has exactly one parent (exclude root and merge commits)
+        // and the commit has not changed the tree content
+        context.parents().len() == 1
```

Comment:
I think we should drop this check, so it works on merge commits too. That will make this `empty` true for most merge commits, which I understand can strange. However, it's more composable that; the condition to filter out merge commits could instead be done in the template (though we may need another template keyword for `merge` for that). We could consider doing that by default, but my inclination is that we should at least try start by just letting merge commits have the "(empty)" label to start with. That might help users get used to the idea that merge commits are typically empty.

We did the same thing for the `empty()` revset (which is just an alias for `~file(".")`) - that revset includes empty merges, so the user has to do `empty() - merges()` if they don't want merges.

I see you're already using `merge_commit_trees()` below, so just removing this condition should be enough.

Reply:
Ok, I'll push a new version right away. Right now it will also consider the root commit empty, is that ok?

Reply:
Yes, I think it makes sense that the root commit is considered empty. That matches what the `empty()` revset does.

Reply:
Would it make sense to specially flag if a merge is _not_ empty to make "evil merge" cases more obvious?

Reply:
Yes, that probably makes sense since merge commits are typically empty. That's actually a good example of how it's better that this `empty` template doesn't treat merges specially, because now we can do something like `if(empty, if(!merge, "", "(NOT empty)"), if(merge, "", "(empty)"))` once we have a `merge` keyword.